### DISCUSSION TITLE
New Component: Date/Time picker

### DISCUSSION THREAD

**Discussion by tminhduc2811** [heroui-inc/heroui#311] [Feature requests]
Hi, I think this is a great project. Having a Date/Time picker component would be really helpful

Something like this https://mui.com/components/pickers/

**Comment by neoromantic** [comment]
Anyone found a good solution? Maybe someone had made a wrapper over NextUI to have a nice date picker?

**Reply by getellez** [reply]
I had to create a custom component using the material ui datepicker and I used the material UI providers only within that custom component to avoid style conflicts between material UI and Tailwind (I'm using tailwind and nextui only, in my project)

**Reply by getellez** [reply]
```
type CustomDatePickerProps = {
  birthDate: Date | null;
  setBirthDate: React.Dispatch<React.SetStateAction<Date | null>>;
};
export const CustomDatePicker: FC<CustomDatePickerProps> = ({ birthDate, setBirthDate }) => {
  const { theme } = useTheme();
  return (
    <DatePickerWrapper>
      <DatePicker
        format='dd/MM/yyyy'
        label='Fecha de nacimiento'
        name='birthDate'
        value={birthDate}
        sx={{
          '& .MuiOutlinedInput-root .MuiOutlinedInput-notchedOutline': {
            borderColor: 'gray'
          },
          '&:hover .MuiOutlinedInput-root .MuiOutlinedInput-notchedOutline': {
            borderColor: 'gray'
          },
          '& .MuiOutlinedInput-root.Mui-focused .MuiOutlinedInput-notchedOutline': {
            borderColor: 'gray'
          }
        }}
        slots={{
          openPickerIcon: () => <HiCalendarDays className={theme === 'dark' ? 'text-white' : 'text-gray'} />
        }}
        slotProps={{
          textField: {
            variant: 'outlined',
            inputProps: {
              style: { color: theme === 'dark' ? 'white' : 'gray' }
            },
            InputLabelProps: { style: { color: theme === 'dark' ? 'white' : 'gray' } }
          }
        }}
        onChange={(newValue) => setBirthDate(newValue)}
      />
    </DatePickerWrapper>
  );
};

```

**Comment by tabuckner** [comment]
Am I crazy, or does this not already exist as `<Input type="date" />`?

**Reply by tabuckner** [reply]
No i'm crazy.

**Reply by jrgarciadev** [reply]
https://nextui.org/docs/components/input#input-types

**Reply by ssijak** [reply]
its not well suported in all browsers, and rendering is different in the ones it is. also not a lot of customisation options

**Comment by hoetmaaiers** [comment]
A Date/Time range selector is not possible with native input-types. This would make a good component.

**Comment by popuguytheparrot** [comment]
https://react-spectrum.adobe.com/react-spectrum/Calendar.html

**Comment by rockse** [comment]
I would love to know if this is coming anytime soon. I am considering nextui to build a dashboard and need a date picker.

**Comment by mauriciabad** [comment]
In Storybook there's an Input with type date and another with time.

https://storybook.nextui.org/?path=/story/components-input--input-types
![rdgs](https://github.com/nextui-org/nextui/assets/12821361/a75755dd-d37d-47a5-bc10-af6b12ffef16)

**Reply by jim-king-2000** [reply]
How to see the source of the storybook?

**Reply by serdaronedio** [reply]
it's a native date-picker, you can read this:

https://dev.to/codeclown/styling-a-native-date-input-into-a-custom-no-library-datepicker-2in

**Reply by vijayganeshpk** [reply]
> How to see the source of the storybook?

@jim-king-2000 check https://github.com/nextui-org/nextui/tree/main/packages/components/input/stories

**Reply by jim-king-2000** [reply]
Great! I'll look into it. Thanks.

**Comment by Lzadhito** [comment]
Creating such component should be great! Would styling the native input (such as -webkit-calendar-picker-indicator) get the work done? or better to create a custom date picker as a base component to be reused by another such as range picker?

**Comment by ssijak** [comment]
This is a must-have for a component library; it's a staple component used in most apps.

**Comment by mxvsh** [comment]
any update on this????

**Comment by J4v4Scr1pt** [comment]
+1

**Comment by rivaslive** [comment]
You can create a custom one, using `react-day-picker` for example.

<img width="579" alt="image" src="https://github.com/nextui-org/nextui/assets/39039038/86af89cf-423b-49e2-8067-110b7db3237f">


```tsx
import { Popover, PopoverContent, PopoverTrigger, Input } from '@nextui-org/react';
import dayjs from 'dayjs';
import { CalendarIcon, X } from 'lucide-react';
import { useEffect, useState } from 'react';
import { DayPicker } from "react-day-picker"
import 'react-day-picker/dist/style.css'

import { cn, formatDate } from '@/lib/utils';
import { InputProps } from '@/lib/ui/input.tsx';
import useToggle from '@/hooks/useToggle.ts';

export function DatePicker({
  placeholder,
  onChange,
  value,
  ...rest
}: Omit<InputProps, 'onChange'> & { onChange?: (val?: string) => void }) {
  const [open, actions] = useToggle();
  const [date, setDate] = useState<string>();

  useEffect(() => {
    typeof value !== 'undefined' && setDate(value);
  }, [value]);

  return (
    <Popover isOpen={open} onOpenChange={actions.setVisible}>
      <PopoverTrigger className="text-left">
        <Input
          {...rest}
          startContent={
            <div
              onClick={rest?.onClick}
              className="absolute flex gap-2 flex-shrink-0 items-center z-10"
            >
              <div>
                <CalendarIcon size="20px" />
              </div>{' '}
              {!date && (
                <span className="text-foreground-500 font-normal w-full text-left text-small truncate">
                  {placeholder}
                </span>
              )}
            </div>
          }
          endContent={
            date && (
              <X
                size="20px"
                className="cursor-pointer"
                onClick={(e) => {
                  e?.stopPropagation();
                  e?.preventDefault();
                  setDate('');
                  onChange?.(undefined);
                }}
              />
            )
          }
          placeholder={placeholder}
          value={formatDate(date, false)}
          classNames={{
            input: cn('text-left ps-[30px!important] pt-5 z-20 h-full'),
            innerWrapper: 'flex'
          }}
        />
      </PopoverTrigger>
      <PopoverContent className="w-auto p-0">
        <DayPicker
          mode="single"
          selected={dayjs(date).toDate()}
          onSelect={(_val) => {
            const _date = dayjs(_val).toISOString();
            setDate(_date);
            onChange?.(_date);
            actions.onHidden();
          }}
          initialFocus
        />
      </PopoverContent>
    </Popover>
  );
}
```

This is just a reference code, you can do it the way you see fit.

**Reply by ssijak** [reply]
Why would we use additional external library when react-aria (which nextui uses) has amazing support for date and time pickers. We just need nextui to implement ui part of it.

**Reply by rivaslive** [reply]
@ssijak I don't know that he uses nextui below and worse even having this application open a long time ago and if Nextui uses react-aria he should have integrated this a long time ago, honestly sometimes it is unfortunate how such a basic component has cost so much to arrive, so my solution is a "worse is nothing" because the best solution is the one that NextUI should offer a long time ago

**Reply by boredmage** [reply]
Hi @rivaslive, can you help with the helper functions and hook as I couldn't recreate them properly.
Thanks
```tsx
import { cn, formatDate } from '@/lib/utils';
import { InputProps } from '@/lib/ui/input.tsx';
import useToggle from '@/hooks/useToggle.ts';
```

**Reply by boredmage** [reply]
Here's my fix.

```tsx
import { InputProps } from "@nextui-org/react";

import { cn, formatDate } from "@/lib/utils";
import useToggle from "@/hooks/useToggle";
```

```ts
// utils.ts
import dayjs from "dayjs";
import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function formatDate(dateString?: string, includeTime = true) {
  if (!dateString) return "";

  return dayjs(dateString).format(
    includeTime ? "DD/MM/YYYY hh:mm A" : "DD/MM/YYYY",
  );
}
```

```ts
// useToggle.ts
import { useState } from "react";

interface ToggleActions {
  setVisible: (visible: boolean) => void;
  onHidden: () => void;
}

function useToggle(initialValue = false): [boolean, ToggleActions] {
  const [isOpen, setIsOpen] = useState(initialValue);

  const toggle = () => setIsOpen(!isOpen);

  const setVisible = (visible: boolean) => setIsOpen(visible);

  const actions = {
    setVisible,
    onHidden: () => setIsOpen(false),
  };

  return [isOpen, actions];
}

export default useToggle;
```

**Reply by jrcopeti** [reply]
@rivaslive  @Qudusayo Thanks for this! I'm using it in my project

**Reply by rivaslive** [reply]
Your welcome!

**Comment by J4v4Scr1pt** [comment]
There is a datepicker in review.😊

https://nextuioss.featurebase.app/p/date-range-picker

**Comment by aprilmintacpineda** [comment]
Why would you guys not want to accept `Date` in `value` prop? 😅

I needed to do this:

```ts
new CalendarDateTime(
  values.datetime.getFullYear(),
  values.datetime.getMonth(),
  values.datetime.getDay(),
  values.datetime.getHours(),
  values.datetime.getMinutes(),
  values.datetime.getSeconds(),
  values.datetime.getMilliseconds()
)
```

Not sure if there's a shorter way of doing this.

**Reply by wingkwong** [reply]
> Why would you guys not want to accept Date in value prop? 😅

By default, JavaScript represents dates and times using the [Date](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) object. However, Date has many problems, including a very difficult to use API, lack of all internationalization support, and more. The [Temporal](https://tc39.es/proposal-temporal/docs/index.html) proposal will eventually address this in the language, and `@internationalized/date` is heavily inspired by it. We hope to back the objects in this package with it once it is implemented in browsers.

> Not sure if there's a shorter way of doing this.

Can you share the full code? Wondering what the type of `values` is and the way you init.

**Comment by boredmage** [comment]
Date component is now available https://nextui.org/blog/v2.3.0
